iceoryx_hoofs/include/iceoryx_hoofs/cxx/expected.hpp

```diff
@@ -95,6 +95,21 @@ struct success<void>
 {
 };
 
+template <typename>
+bool always_false_v = false;
+
+template <typename F, typename T>
+constexpr T from(const F)
+{
+    static_assert(always_false_v<F> && always_false_v<T>, "Conversion for the specified types is not implemented!");
+}
+
+template <typename T, typename F>
+constexpr T into(const F e)
+{
+    return from<F, T>(e);
+}
+
```

Comment:
This will not be merged with this PR. It is just here to show you a concept from Rust which is also used for error handling. There are the `From` and `Into` traits and when `From` is implemented, `Into` is automatically derived. With the `cxx::expected` we often have to pass an error through multiple layers. With this free functions we can create a contract to automate this. One just has to implement
```cpp
template <>
constexpr Bar from<Foo, Bar>(Foo f) {...}
```
and could then just do a `return cxx::error<Bar>::from(expectedWithFooError.get_error());` (note, this `from` is a different one than the one from above, but calls the one from above)

Additionally one can just call `auto barError = into<Bar>(fooError);`

If you agree that we should discuss this further, I will create an issue for this.

@elfenpiff @dkroenke @FerdinandSpitzschnueffler @mossmaurice @MatthiasKillat

Reply:
The part with the conversion is done with #992.
Do we also want `cxx::error<T>::from(...)` static function?

Reply:
The `AllocationError` has errors like `TOO_MANY_CHUNKS_ALLOCATED_IN_PARALLEL` which are not relevant for the `MemoryManager`. When we use one enum for all, we have to check for this error in all places and the user also has to handle this errors even if we condense multiple `MemoryManager::Error` to a single `ApplicationError`. 

The `from`/`into` is just an internal detail to make our live easier and create a consistent way to convert between enums